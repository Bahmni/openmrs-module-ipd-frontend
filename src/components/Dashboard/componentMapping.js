import { lazy } from "react";

export const componentMapping = {
  PS: lazy(() =>
    import("../../features/DisplayControls/PatientSummary/PatientSummary")
  ),
  VT: lazy(() => import("../../features/DisplayControls/Vitals/Vitals")),
  TR: lazy(() =>
    import("../../features/DisplayControls/Treatments/components/Treatments")
  ),
  AL: lazy(() =>
    import("../../features/DisplayControls/Allergies/components/Allergies")
  ),
};
