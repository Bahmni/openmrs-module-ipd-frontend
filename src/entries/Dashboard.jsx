import React from "react";
import PropTypes from "prop-types";
import "./Dashboard.scss";
import { I18nProvider } from "../features/i18n/I18nProvider";
import DrugChartWrapper from "../features/DrugChart/DrugChartWrapper/DrugChartWrapper.jsx";
/**
 * NOTE: Currently there is nothing in this dashboard, all debugging items
 */
export default function Dashboard(props) {
  const { hostData } = props;
  return (
    <>
      <I18nProvider>
        {/* <Button
            kind="primary"
            onClick={() => hostApi.onConfirm?.("event-from-ipd")}
          >
            Click to send event to host appliation
          </Button> */}
        <DrugChartWrapper
          patientId={hostData?.patientId}
          viewDate={hostData?.forDate}
        />
      </I18nProvider>
    </>
  );
}
Dashboard.propTypes = {
  hostData: PropTypes.shape({
    patientId: PropTypes.string,
    forDate: PropTypes.instanceOf(Date),
  }).isRequired,
};
