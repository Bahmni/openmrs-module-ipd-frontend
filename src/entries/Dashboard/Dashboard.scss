.ipd-page {
  .main {
    transition: margin-left 0.5s;
    padding: 16px;
    position: relative;
    margin-left: 16rem;
  }

  .main-with-slider {
    transition: margin-left 0.5s;
    padding: 16px;
    position: relative;
    // width: calc(100% - 550px - 16rem);
    margin-left: 16rem;
    margin-right: 540px;
  }

  .bx--accordion__title {
    font-weight: 600;
  }

  .bx--side-nav__overlay-active {
    width: 0;
    height: 0;
    background-color: rgba(0, 0, 0, 0);
    opacity: 0;
    transition: opacity 240ms cubic-bezier(0.2, 0, 0.38, 0.9),
      background-color 240ms cubic-bezier(0.2, 0, 0.38, 0.9);
  }

  .border-bottom-0 {
    border-bottom: 0;
  }

  .navbar-border {
    border-right: 1px solid #007d79;
  }

  .cursor-pointer {
    cursor: pointer;
  }

  a.bx--side-nav__link--current::before {
    background-color: #007d79;
  }

  .header-bg-color {
    background: #004144;
  }

  .header-nav-toggle-btn {
    border: none;
  }

  .header-nav-toggle-btn:hover {
    background-color: #004144;
    fill: #ffffff;
    outline: none;
    border: none;
  }

  .bx--accordion__heading {
    border-bottom: 1px solid #007d79;
    border-radius: 0;
  }

  .bx--accordion__item:last-child {
    border: none;
  }
  svg[class*="bx--accordion__arrow"] {
    fill: #666;
  }
  .bx--accordion__content {
    padding: 0;
  }

  .navigation-buttons {
    color: #0f62fe;
    display: flex;
    justify-content: space-between;
    margin-bottom: 5px;
    > * {
      cursor: pointer;
    }
  }
}

@media (max-width: 1024px) {
  .ipd-page {
    .main {
      margin-left: 0;
    }
    .main-with-slider {
      margin-left: 0;
      // width: calc(100% - 550px);
      margin-right: 540px;
    }
  }
}

@media (max-width: 640px) {
  .header-nav-toggle-btn {
    display: none;
  }
}
