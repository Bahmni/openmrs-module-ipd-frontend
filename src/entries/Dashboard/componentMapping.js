import { lazy } from "react";
import { componentKeys } from "../../constants";

export const componentMapping = {
  [componentKeys.DIAGNOSIS]: lazy(() =>
    import("../../features/DisplayControls/Diagnosis/components/Diagnosis")
  ),
  [componentKeys.TREATMENTS]: lazy(() =>
    import("../../features/DisplayControls/Treatments/components/Treatments")
  ),
  [componentKeys.ALLERGIES]: lazy(() =>
    import("../../features/DisplayControls/Allergies/components/Allergies")
  ),
  [componentKeys.NURSING_TASKS]: lazy(() =>
    import(
      "../../features/DisplayControls/NursingTasks/components/NursingTasks"
    )
  ),
};
