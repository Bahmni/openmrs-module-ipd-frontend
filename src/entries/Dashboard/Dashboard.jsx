import React, { useState, useRef, Suspense, useEffect } from "react";
import { FormattedMessage } from "react-intl";
import {
  Accordion,
  AccordionItem,
  Header,
  HeaderMenuButton,
  SideNav,
  SideNavItems,
  SideNavLink,
  Link,
} from "carbon-components-react";
import { ArrowLeft } from "@carbon/icons-react/next";
import { componentMapping } from "./componentMapping";
import "./Dashboard.scss";
import data from "../../utils/config.json";
import PropTypes from "prop-types";
import { I18nProvider } from "../../features/i18n/I18nProvider";
import { getPatientDashboardUrl } from "../../utils/CommonUtils";
import { PatientHeader } from "../../features/DisplayControls/PatientHeader/components/PatientHeader";
import RefreshDisplayControl from "../../context/RefreshDisplayControl";
import { SliderContext } from "../../context/SliderContext";

export default function Dashboard(props) {
  const { hostData, hostApi } = props;
  const { patient } = hostData;
  const [sliderContentModified, setSliderContentModified] = useState({
    treatments: false,
    nursingTasks: false,
    emergencyTasks: false,
  });
  const [isSliderOpen, updateSliderOpen] = useState({
    treatments: false,
    nursingTasks: false,
    emergencyTasks: false,
  });
  const [sections, setSections] = useState([]);
  const [isSideNavExpanded, updateSideNav] = useState(true);
  const [selectedTab, updateSelectedTab] = useState(null);
  const refs = useRef([]);
  const [windowWidth, updateWindowWidth] = useState(window.outerWidth);

  window.addEventListener("resize", () => {
    updateWindowWidth(window.outerWidth);
  });
  useEffect(() => {
    updateSideNav(window.outerWidth > 1024);
  }, [windowWidth]);

  const fetchConfig = () => {
    const { config: { sections = [] } = {} } = data;
    const updatedSections = sections
      .filter((sec) => componentMapping[sec.componentKey])
      .sort((a, b) => a.displayOrder - b.displayOrder);
    updateSelectedTab(updatedSections[0].componentKey);
    setSections(updatedSections);
  };

  useEffect(() => {
    fetchConfig();
  }, []);

  const onClickSideNavExpand = () => {
    updateSideNav((oldState) => !oldState);
  };

  const scrollToSection = (key) => {
    updateSelectedTab(key);
    window.scrollTo({
      top: refs.current[key].offsetTop,
      behavior: "smooth",
    });
  };

  const checkSliderStatus = () => {
    return Object.values(isSliderOpen).includes(true);
  };

  const handleDischargeSummaryNavigation = () => {
    hostApi?.navigation?.dischargeSummary?.();
  };

  const refreshDisplayControl = (componentKeyArray) => {
    const updatedSections = sections.map((el) => {
      return componentKeyArray.includes(el.componentKey)
        ? { ...el, refreshKey: Math.random() }
        : el;
    });
    setSections(updatedSections);
  };

  return (
    <SliderContext.Provider
      value={{
        isSliderOpen,
        updateSliderOpen,
        sliderContentModified,
        setSliderContentModified,
      }}
    >
      <main className="ipd-page">
        <Header
          className="border-bottom-0 header-bg-color"
          aria-label="IBM Platform Name"
        >
          <HeaderMenuButton
            aria-label="Open menu"
            className="header-nav-toggle-btn"
            onClick={onClickSideNavExpand}
            isActive={isSideNavExpanded}
          />
          <SideNav
            aria-label="Side navigation"
            className="navbar-border"
            isPersistent={true}
            expanded={isSideNavExpanded}
          >
            <SideNavItems>
              {sections?.map((el) => {
                return (
                  <SideNavLink
                    className="cursor-pointer"
                    isActive={el.componentKey === selectedTab}
                    key={el.componentKey}
                    onClick={() => scrollToSection(el.componentKey)}
                  >
                    {el.title}
                  </SideNavLink>
                );
              })}
            </SideNavItems>
          </SideNav>
        </Header>

        <section className={checkSliderStatus() ? "main-with-slider" : "main"}>
          <div className={"navigation-buttons"}>
            <ArrowLeft
              data-testid={"Back button"}
              size={20}
              onClick={() => window.history.back()}
            />
            <Link
              onClick={() => {
                window.location.href = getPatientDashboardUrl(patient?.uuid);
              }}
            >
              <FormattedMessage
                id={"VIEW_CLINICAL_DASHBOARD"}
                defaultMessage={"View Clinical Dashboard"}
              />
            </Link>
          </div>
          <PatientHeader
            patientId={patient?.uuid}
            openDischargeSummary={handleDischargeSummaryNavigation}
          />
          <Accordion className={"accordion"}>
            {sections?.map((el) => {
              const DisplayControl = componentMapping[el.componentKey];
              return (
                <section
                  key={el.componentKey}
                  ref={(ref) => (refs.current[el.componentKey] = ref)}
                  style={{ marginBottom: "40px" }}
                >
                  <Suspense fallback={<p>Loading...</p>}>
                    <AccordionItem open title={el.title}>
                      <I18nProvider>
                        <RefreshDisplayControl.Provider
                          value={refreshDisplayControl}
                        >
                          <DisplayControl
                            key={el.refreshKey}
                            patientId={patient?.uuid}
                          />
                        </RefreshDisplayControl.Provider>
                      </I18nProvider>
                    </AccordionItem>
                  </Suspense>
                </section>
              );
            })}
          </Accordion>
        </section>
      </main>
    </SliderContext.Provider>
  );
}
Dashboard.propTypes = {
  hostData: PropTypes.object.isRequired,
  hostApi: PropTypes.object.isRequired,
};
