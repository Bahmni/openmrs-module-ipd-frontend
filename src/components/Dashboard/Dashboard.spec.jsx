import React from "react";
import { act, render, fireEvent, screen } from "@testing-library/react";

import Dashboard from "./Dashboard";

describe("Dashboard", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<Dashboard />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders the header", () => {
    render(<Dashboard />);
    const header = screen.getByRole("banner");
    expect(header).toBeTruthy();
  });

  it("renders the side navigation", () => {
    render(<Dashboard />);
    const sideNav = screen.getByRole("navigation");
    expect(sideNav).toBeTruthy();
  });

  it("renders the accordion sections", () => {
    render(<Dashboard />);
    const section1 = screen.findAllByText("Patient Summary");
    const section2 = screen.findAllByText("Vitals");
    expect(section1).toBeTruthy();
    expect(section2).toBeTruthy();
  });

  it("should update window width on resize", () => {
    render(<Dashboard />);

    expect(window.innerWidth).toEqual(window.outerWidth);

    act(() => {
      window.innerWidth = 1200;
      window.dispatchEvent(new Event("resize"));
    });

    render(<Dashboard />);

    expect(window.innerWidth).toEqual(1200);
  });

  it("should update sidenav on click of side nav expand", async () => {
    render(<Dashboard />);

    const menuButton = screen.getByLabelText("Open menu");

    fireEvent.click(menuButton);
    expect(screen.getByRole("navigation")).toBeTruthy();
  });
});
