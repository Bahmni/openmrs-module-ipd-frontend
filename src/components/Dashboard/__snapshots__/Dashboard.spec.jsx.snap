// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match snapshot 1`] = `
<DocumentFragment>
  <main
    class="ipd-page"
  >
    <header
      aria-label="IBM Platform Name"
      class="bx--header border-bottom-0 header-bg-color"
    >
      <button
        aria-label="Open menu"
        class="header-nav-toggle-btn bx--header__action bx--header__menu-trigger bx--header__menu-toggle"
        title="Open menu"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="20"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 14.8H18V16H2zM2 11.2H18V12.399999999999999H2zM2 7.6H18V8.799999999999999H2zM2 4H18V5.2H2z"
          />
        </svg>
      </button>
      <div
        class="bx--side-nav__overlay"
      />
      <nav
        aria-label="Side navigation"
        class="bx--side-nav__navigation bx--side-nav navbar-border bx--side-nav--ux bx--side-nav--hidden"
      >
        <ul
          class="bx--side-nav__items"
        >
          <li
            class="bx--side-nav__item"
          >
            <a
              class="bx--side-nav__link bx--side-nav__link--current cursor-pointer"
            >
              <span
                class="bx--side-nav__link-text"
              >
                Patient Summary
              </span>
            </a>
          </li>
          <li
            class="bx--side-nav__item"
          >
            <a
              class="bx--side-nav__link cursor-pointer"
            >
              <span
                class="bx--side-nav__link-text"
              >
                Vitals
              </span>
            </a>
          </li>
          <li
            class="bx--side-nav__item"
          >
            <a
              class="bx--side-nav__link cursor-pointer"
            >
              <span
                class="bx--side-nav__link-text"
              >
                Allergies
              </span>
            </a>
          </li>
          <li
            class="bx--side-nav__item"
          >
            <a
              class="bx--side-nav__link cursor-pointer"
            >
              <span
                class="bx--side-nav__link-text"
              >
                Treatments
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </header>
    <section
      class="main"
    >
      <ul
        class="bx--accordion accordion bx--accordion--end bx--accordion--md"
      >
        <section
          style="margin-bottom: 40px;"
        >
          <li
            class="bx--accordion__item bx--accordion__item--active"
          >
            <button
              aria-controls="accordion-item-1"
              aria-expanded="true"
              class="bx--accordion__heading"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
              <div
                class="bx--accordion__title"
                dir="auto"
              >
                Patient Summary
              </div>
            </button>
            <div
              class="bx--accordion__content"
              id="accordion-item-1"
            >
              <div />
            </div>
          </li>
        </section>
        <section
          style="margin-bottom: 40px;"
        >
          <li
            class="bx--accordion__item bx--accordion__item--active"
          >
            <button
              aria-controls="accordion-item-2"
              aria-expanded="true"
              class="bx--accordion__heading"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
              <div
                class="bx--accordion__title"
                dir="auto"
              >
                Vitals
              </div>
            </button>
            <div
              class="bx--accordion__content"
              id="accordion-item-2"
            >
              <div />
            </div>
          </li>
        </section>
        <section
          style="margin-bottom: 40px;"
        >
          <li
            class="bx--accordion__item bx--accordion__item--active"
          >
            <button
              aria-controls="accordion-item-3"
              aria-expanded="true"
              class="bx--accordion__heading"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
              <div
                class="bx--accordion__title"
                dir="auto"
              >
                Allergies
              </div>
            </button>
            <div
              class="bx--accordion__content"
              id="accordion-item-3"
            >
              <div />
            </div>
          </li>
        </section>
        <section
          style="margin-bottom: 40px;"
        >
          <li
            class="bx--accordion__item bx--accordion__item--active"
          >
            <button
              aria-controls="accordion-item-4"
              aria-expanded="true"
              class="bx--accordion__heading"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
              <div
                class="bx--accordion__title"
                dir="auto"
              >
                Treatments
              </div>
            </button>
            <div
              class="bx--accordion__content"
              id="accordion-item-4"
            >
              <div />
            </div>
          </li>
        </section>
      </ul>
    </section>
  </main>
</DocumentFragment>
`;
